File: tape/sdk/typescript/src/chaos/proxies.ts
// Agent-layer chaos proxies — Toxiproxy-for-LLM and -MCP.
//
// Mirrors `tape.chaos.proxies` from the Python SDK. Stdlib-only (`node:http`,
// `node:https`) — no extra dependency. The proxy is plaintext-to-agent;
// it upgrades to TLS on the way out if the upstream is `https://`.
//
//   import * as chaos from 'tape-ts/chaos';
//
//   const p = chaos.modelProxy('https://api.anthropic.com', [
//     chaos.proxy.delay({ pathPrefix: '/v1/messages', ms: 2_000, probability: 0.1 }),
//     chaos.proxy.injectStatus({ pathPrefix: '/v1/messages', status: 429, probability: 0.05 }),
//   ]);
//   await p.start();   // listens on 127.0.0.1:auto-port
//   // point your LLM client's baseUrl at p.url
//   await p.stop();

import * as http from 'node:http';
import * as https from 'node:https';
import { URL } from 'node:url';
import type { AddressInfo } from 'node:net';

export type FaultKind =
  | 'delay'
  | 'inject_status'
  | 'truncate_stream'
  | 'mangle_json'
  | 'inject_prompt'
  | 'tool_shadow'
  | 'schema_drift'
  | 'drop_connection';

export interface ProxyFault {
  readonly kind: FaultKind;
  readonly pathPrefix: string;
  readonly probability: number;
  readonly ms: number;
  readonly jitter: number;
  readonly status: number;
  readonly body: string;
  readonly atEvent: number;
  readonly jsonPath: string;
  readonly replacement: unknown;
  readonly suffix: string;
  readonly extraTool: Record<string, unknown> | null;
  readonly driftFn: ((payload: unknown) => unknown) | null;
}

function fault(partial: Partial<ProxyFault> & { kind: FaultKind }): ProxyFault {
  return {
    pathPrefix: '', probability: 1.0, ms: 0, jitter: 0,
    status: 0, body: '', atEvent: 0, jsonPath: '',
    replacement: null, suffix: '', extraTool: null, driftFn: null,
    ...partial,
  };
}

// ── fault constructors ─────────────────────────────────────────────────────

export function delay(opts: { pathPrefix?: string; ms: number; jitter?: number; probability?: number }): ProxyFault {
  return fault({ kind: 'delay', pathPrefix: opts.pathPrefix ?? '', ms: opts.ms,
                  jitter: opts.jitter ?? 0, probability: opts.probability ?? 1.0 });
}

export function injectStatus(opts: { pathPrefix?: string; status: number; body?: string; probability?: number }): ProxyFault {
  return fault({ kind: 'inject_status', pathPrefix: opts.pathPrefix ?? '',
                  status: opts.status, body: opts.body ?? '',
                  probability: opts.probability ?? 1.0 });
}

export function truncateStream(opts: { pathPrefix?: string; atEvent: number; probability?: number }): ProxyFault {
  return fault({ kind: 'truncate_stream', pathPrefix: opts.pathPrefix ?? '',
                  atEvent: opts.atEvent, probability: opts.probability ?? 1.0 });
}

export function mangleJson(opts: { pathPrefix?: string; jsonPath?: string; replacement?: unknown; probability?: number }): ProxyFault {
  return fault({ kind: 'mangle_json', pathPrefix: opts.pathPrefix ?? '',
                  jsonPath: opts.jsonPath ?? '', replacement: opts.replacement ?? null,
                  probability: opts.probability ?? 1.0 });
}

export function injectPrompt(opts: { pathPrefix?: string; suffix: string; probability?: number }): ProxyFault {
  return fault({ kind: 'inject_prompt', pathPrefix: opts.pathPrefix ?? '',
                  suffix: opts.suffix, probability: opts.probability ?? 1.0 });
}

export function toolShadow(opts: { pathPrefix?: string; extraTool: Record<string, unknown>; probability?: number }): ProxyFault {
  return fault({ kind: 'tool_shadow', pathPrefix: opts.pathPrefix ?? '',
                  extraTool: opts.extraTool, probability: opts.probability ?? 1.0 });
}

export function schemaDrift(opts: { pathPrefix?: string; driftFn: (payload: unknown) => unknown; probability?: number }): ProxyFault {
  return fault({ kind: 'schema_drift', pathPrefix: opts.pathPrefix ?? '',
                  driftFn: opts.driftFn, probability: opts.probability ?? 1.0 });
}

export function dropConnection(opts: { pathPrefix?: string; probability?: number } = {}): ProxyFault {
  return fault({ kind: 'drop_connection', pathPrefix: opts.pathPrefix ?? '',
                  probability: opts.probability ?? 1.0 });
}

// ── helpers ────────────────────────────────────────────────────────────────

function setJsonAt(obj: any, path: string, value: unknown): any {
  if (!path) return value;
  const parts = path.split('.');
  let cur: any = obj;
  for (let i = 0; i < parts.length - 1; i++) {
    const p = parts[i];
    try {
      if (Array.isArray(cur) && /^\d+$/.test(p)) cur = cur[Number(p)];
      else if (cur && typeof cur === 'object') cur = cur[p];
      else return obj;
    } catch { return obj; }
  }
  const last = parts[parts.length - 1];
  try {
    if (Array.isArray(cur) && /^\d+$/.test(last)) cur[Number(last)] = value;
    else if (cur && typeof cur === 'object') cur[last] = value;
  } catch { /* ignore */ }
  return obj;
}

function injectPromptInto(obj: any, suffix: string): any {
  if (Array.isArray(obj)) for (const v of obj) injectPromptInto(v, suffix);
  else if (obj && typeof obj === 'object') {
    for (const k of ['text', 'content', 'output_text']) {
      if (typeof obj[k] === 'string') obj[k] = obj[k] + suffix;
    }
    for (const v of Object.values(obj)) injectPromptInto(v, suffix);
  }
  return obj;
}

function shadowTools(obj: any, extra: Record<string, unknown>): any {
  if (Array.isArray(obj)) for (const v of obj) shadowTools(v, extra);
  else if (obj && typeof obj === 'object') {
    for (const [k, v] of Object.entries(obj)) {
      if (k === 'tools' && Array.isArray(v)) (v as unknown[]).push({ ...extra });
      else shadowTools(v, extra);
    }
  }
  return obj;
}

const DROP_REQUEST_HEADERS = new Set([
  'host', 'content-length', 'connection', 'keep-alive',
  'proxy-authenticate', 'proxy-authorization', 'te', 'trailers',
  'transfer-encoding', 'upgrade',
]);

// ── ChaosProxy ─────────────────────────────────────────────────────────────

export class ChaosProxy {
  readonly upstream: string;
  readonly faults: readonly ProxyFault[];
  private readonly rng: () => number;
  private readonly timeoutMs: number;
  private server: http.Server | null = null;
  private _host = '';
  private _port = 0;
  readonly faultHits = new Map<string, number>();

  constructor(upstream: string, faults: readonly ProxyFault[] = [],
              opts: { rng?: () => number; timeoutMs?: number } = {}) {
    this.upstream = upstream.replace(/\/$/, '');
    this.faults = faults;
    this.rng = opts.rng ?? Math.random;
    this.timeoutMs = opts.timeoutMs ?? 60_000;
  }

  get url(): string {
    return `http://${this._host}:${this._port}`;
  }

  async start(opts: { host?: string; port?: number } = {}): Promise<void> {
    const server = http.createServer((req, res) => this.handle(req, res));
    await new Promise<void>((resolve) => {
      server.listen(opts.port ?? 0, opts.host ?? '127.0.0.1', () => resolve());
    });
    const addr = server.address() as AddressInfo;
    this._host = addr.address; this._port = addr.port;
    this.server = server;
  }

  async stop(): Promise<void> {
    if (!this.server) return;
    await new Promise<void>((resolve, reject) => {
      this.server!.close((err) => err ? reject(err) : resolve());
    });
    this.server = null;
  }

  private matching(path: string, kind: FaultKind): ProxyFault[] {
    const out: ProxyFault[] = [];
    for (const f of this.faults) {
      if (f.kind !== kind) continue;
      if (f.pathPrefix && !path.startsWith(f.pathPrefix)) continue;
      if (f.probability >= 1.0 || this.rng() < f.probability) {
        out.push(f);
        const key = `${kind}:${f.pathPrefix}`;
        this.faultHits.set(key, (this.faultHits.get(key) ?? 0) + 1);
      }
    }
    return out;
  }

  private async handle(req: http.IncomingMessage, res: http.ServerResponse): Promise<void> {
    const path = req.url ?? '/';

    // PRE-FORWARD faults: delay, inject_status.
    for (const f of this.matching(path, 'delay')) {
      let ms = f.ms;
      if (f.jitter > 0) ms = Math.max(0, Math.floor(ms * (1.0 + (this.rng() * 2 - 1) * f.jitter)));
      await new Promise(r => setTimeout(r, ms));
    }
    const injected = this.matching(path, 'inject_status');
    if (injected.length) {
      const f = injected[0];
      const body = Buffer.from(f.body, 'utf-8');
      res.writeHead(f.status, {
        'Content-Type': 'text/plain; charset=utf-8',
        'Content-Length': body.length,
        'X-Tape-Chaos': 'inject_status',
      });
      res.end(body);
      return;
    }

    // Read the request body fully (proxies are lower-throughput; OK).
    const chunks: Buffer[] = [];
    for await (const ch of req) chunks.push(ch as Buffer);
    const body = Buffer.concat(chunks);

    // Forward.
    const target = new URL(this.upstream + path);
    const lib = target.protocol === 'https:' ? https : http;
    const upstreamHeaders: Record<string, string> = {};
    for (const [k, v] of Object.entries(req.headers)) {
      if (DROP_REQUEST_HEADERS.has(k.toLowerCase())) continue;
      if (v === undefined) continue;
      upstreamHeaders[k] = Array.isArray(v) ? v.join(',') : String(v);
    }
    if (body.length) upstreamHeaders['content-length'] = String(body.length);

    let upstreamRes: http.IncomingMessage;
    try {
      upstreamRes = await new Promise<http.IncomingMessage>((resolve, reject) => {
        const r = lib.request({
          method: req.method, hostname: target.hostname,
          port: target.port || (target.protocol === 'https:' ? 443 : 80),
          path: target.pathname + target.search,
          headers: upstreamHeaders, timeout: this.timeoutMs,
        }, resolve);
        r.on('error', reject);
        r.on('timeout', () => { r.destroy(new Error('upstream timeout')); });
        if (body.length) r.write(body);
        r.end();
      });
    } catch (ex) {
      res.writeHead(502, { 'X-Tape-Chaos': 'upstream-unreachable' });
      res.end(`upstream unreachable: ${ex instanceof Error ? ex.message : String(ex)}`);
      return;
    }

    const ctype = (upstreamRes.headers['content-type'] ?? '').toString();
    if (ctype.startsWith('text/event-stream')) {
      await this.replyStream(path, upstreamRes, res);
    } else if (ctype.startsWith('application/json')) {
      await this.replyJson(path, upstreamRes, res);
    } else {
      await this.replyPassthrough(path, upstreamRes, res);
    }
  }

  private sendHeaders(upstreamRes: http.IncomingMessage, res: http.ServerResponse,
                       opts: { overrideLength?: number; extra?: Record<string, string> } = {}): void {
    const headers: Record<string, string | number> = {};
    for (const [k, v] of Object.entries(upstreamRes.headers)) {
      const lk = k.toLowerCase();
      if (lk === 'transfer-encoding' || lk === 'content-encoding'
          || lk === 'connection' || lk === 'keep-alive') continue;
      if (opts.overrideLength !== undefined && lk === 'content-length') continue;
      if (v !== undefined) headers[k] = Array.isArray(v) ? v.join(',') : String(v);
    }
    if (opts.overrideLength !== undefined) headers['Content-Length'] = opts.overrideLength;
    if (opts.extra) Object.assign(headers, opts.extra);
    res.writeHead(upstreamRes.statusCode ?? 200, headers);
  }

  private async replyPassthrough(path: string, upstreamRes: http.IncomingMessage, res: http.ServerResponse): Promise<void> {
    const chunks: Buffer[] = [];
    for await (const ch of upstreamRes) chunks.push(ch as Buffer);
    const data = Buffer.concat(chunks);
    const drops = this.matching(path, 'drop_connection');
    this.sendHeaders(upstreamRes, res, { overrideLength: data.length, extra: { 'X-Tape-Chaos': 'passthrough' } });
    if (drops.length) { res.end(); return; }
    res.end(data);
  }

  private async replyJson(path: string, upstreamRes: http.IncomingMessage, res: http.ServerResponse): Promise<void> {
    const chunks: Buffer[] = [];
    for await (const ch of upstreamRes) chunks.push(ch as Buffer);
    const data = Buffer.concat(chunks);
    let payload: any;
    try { payload = JSON.parse(data.toString('utf-8')); }
    catch {
      this.sendHeaders(upstreamRes, res, { overrideLength: data.length });
      res.end(data); return;
    }
    const applied: string[] = [];
    for (const f of this.matching(path, 'mangle_json')) {
      payload = setJsonAt(payload, f.jsonPath, f.replacement);
      applied.push('mangle_json');
    }
    for (const f of this.matching(path, 'inject_prompt')) {
      payload = injectPromptInto(payload, f.suffix); applied.push('inject_prompt');
    }
    for (const f of this.matching(path, 'tool_shadow')) {
      if (f.extraTool) { payload = shadowTools(payload, { ...f.extraTool }); applied.push('tool_shadow'); }
    }
    for (const f of this.matching(path, 'schema_drift')) {
      if (f.driftFn) {
        try { payload = f.driftFn(payload); }
        catch (ex) { console.warn('proxy: schemaDrift threw:', ex); }
        applied.push('schema_drift');
      }
    }
    const drops = this.matching(path, 'drop_connection');
    const newBody = Buffer.from(JSON.stringify(payload), 'utf-8');
    this.sendHeaders(upstreamRes, res, {
      overrideLength: newBody.length,
      extra: { 'X-Tape-Chaos': applied.join(',') || 'json' },
    });
    if (drops.length) { res.end(); return; }
    res.end(newBody);
  }

  private async replyStream(path: string, upstreamRes: http.IncomingMessage, res: http.ServerResponse): Promise<void> {
    const truncates = this.matching(path, 'truncate_stream');
    const cutAt = Math.min(...(truncates.length ? truncates.map(t => t.atEvent) : [0]));
    const drops = this.matching(path, 'drop_connection');
    this.sendHeaders(upstreamRes, res, {
      extra: { 'X-Tape-Chaos': truncates.length ? 'truncate_stream'
                                : drops.length ? 'drop_connection' : 'sse' },
    });
    let eventCount = 0;
    let buf = Buffer.alloc(0);
    for await (const chunk of upstreamRes) {
      buf = Buffer.concat([buf, chunk as Buffer]);
      while (true) {
        const idx = buf.indexOf('\n\n');
        if (idx < 0) break;
        const evt = buf.slice(0, idx + 2);
        buf = buf.slice(idx + 2);
        eventCount++;
        if (!res.write(evt)) await new Promise<void>(r => res.once('drain', r));
        if (cutAt && eventCount >= cutAt) { res.end(); return; }
      }
      if (drops.length && eventCount >= 1) { res.end(); return; }
    }
    res.end();
  }
}

// ── convenience constructors ───────────────────────────────────────────────

/** A `ChaosProxy` tuned for an LLM provider's `baseUrl` (Anthropic/OpenAI/Vertex). */
export function modelProxy(upstream: string, faults: readonly ProxyFault[] = [],
                            opts: { rng?: () => number; timeoutMs?: number } = {}): ChaosProxy {
  return new ChaosProxy(upstream, faults, opts);
}

/** A `ChaosProxy` tuned for an MCP server's HTTP/SSE endpoint. */
export function mcpProxy(upstream: string, faults: readonly ProxyFault[] = [],
                          opts: { rng?: () => number; timeoutMs?: number } = {}): ChaosProxy {
  return new ChaosProxy(upstream, faults, opts);
}

File: tape/sdk/go/chaos/proxies.go
package chaos

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"math/rand/v2"
	"net"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"sync"
	"time"
)

// ProxyFaultKind — the eight v1 fault kinds the proxy honours.
type ProxyFaultKind string

const (
	FaultDelay          ProxyFaultKind = "delay"
	FaultInjectStatus   ProxyFaultKind = "inject_status"
	FaultTruncateStream ProxyFaultKind = "truncate_stream"
	FaultMangleJSON     ProxyFaultKind = "mangle_json"
	FaultInjectPrompt   ProxyFaultKind = "inject_prompt"
	FaultToolShadow     ProxyFaultKind = "tool_shadow"
	FaultSchemaDrift    ProxyFaultKind = "schema_drift"
	FaultDropConnection ProxyFaultKind = "drop_connection"
)

// ProxyFault — one declarative chaos rule the ChaosProxy applies.
type ProxyFault struct {
	Kind        ProxyFaultKind
	PathPrefix  string
	Probability float64

	Ms      int
	Jitter  float64
	Status  int
	Body    string
	AtEvent int

	JSONPath    string
	Replacement any
	Suffix      string
	ExtraTool   map[string]any
	DriftFn     func(payload any) any
}

// ── fault constructors ───────────────────────────────────────────────────

// PDelay — sleep ms ± jitter before forwarding.
func PDelay(pathPrefix string, ms int, probability float64) ProxyFault {
	return ProxyFault{Kind: FaultDelay, PathPrefix: pathPrefix, Ms: ms, Probability: probability}
}

// PInjectStatus — short-circuit with HTTP status + body.
func PInjectStatus(pathPrefix string, status int, body string, probability float64) ProxyFault {
	return ProxyFault{Kind: FaultInjectStatus, PathPrefix: pathPrefix, Status: status, Body: body, Probability: probability}
}

// PTruncateStream — cut SSE after `atEvent` events.
func PTruncateStream(pathPrefix string, atEvent int, probability float64) ProxyFault {
	return ProxyFault{Kind: FaultTruncateStream, PathPrefix: pathPrefix, AtEvent: atEvent, Probability: probability}
}

// PMangleJSON — replace a dotted JSON field on the response.
func PMangleJSON(pathPrefix, jsonPath string, replacement any, probability float64) ProxyFault {
	return ProxyFault{Kind: FaultMangleJSON, PathPrefix: pathPrefix, JSONPath: jsonPath, Replacement: replacement, Probability: probability}
}

// PInjectPrompt — append a suffix to top-level text/content fields.
func PInjectPrompt(pathPrefix, suffix string, probability float64) ProxyFault {
	return ProxyFault{Kind: FaultInjectPrompt, PathPrefix: pathPrefix, Suffix: suffix, Probability: probability}
}

// PToolShadow — inject an extra tool into an MCP tools/list response.
func PToolShadow(pathPrefix string, extraTool map[string]any, probability float64) ProxyFault {
	return ProxyFault{Kind: FaultToolShadow, PathPrefix: pathPrefix, ExtraTool: extraTool, Probability: probability}
}

// PSchemaDrift — apply an arbitrary transform (escape hatch).
func PSchemaDrift(pathPrefix string, driftFn func(any) any, probability float64) ProxyFault {
	return ProxyFault{Kind: FaultSchemaDrift, PathPrefix: pathPrefix, DriftFn: driftFn, Probability: probability}
}

// PDropConnection — close the socket mid-response.
func PDropConnection(pathPrefix string, probability float64) ProxyFault {
	return ProxyFault{Kind: FaultDropConnection, PathPrefix: pathPrefix, Probability: probability}
}

// ── helpers ──────────────────────────────────────────────────────────────

func setJSONAt(obj any, path string, value any) any {
	if path == "" {
		return value
	}
	parts := strings.Split(path, ".")
	cur := obj
	for i := 0; i < len(parts)-1; i++ {
		p := parts[i]
		if arr, ok := cur.([]any); ok {
			n, err := strconv.Atoi(p)
			if err != nil || n < 0 || n >= len(arr) {
				return obj
			}
			cur = arr[n]
		} else if m, ok := cur.(map[string]any); ok {
			cur = m[p]
		} else {
			return obj
		}
	}
	last := parts[len(parts)-1]
	if arr, ok := cur.([]any); ok {
		n, err := strconv.Atoi(last)
		if err == nil && n >= 0 && n < len(arr) {
			arr[n] = value
		}
	} else if m, ok := cur.(map[string]any); ok {
		m[last] = value
	}
	return obj
}

func injectPromptInto(obj any, suffix string) any {
	switch v := obj.(type) {
	case map[string]any:
		for _, k := range []string{"text", "content", "output_text"} {
			if s, ok := v[k].(string); ok {
				v[k] = s + suffix
			}
		}
		for _, vv := range v {
			injectPromptInto(vv, suffix)
		}
	case []any:
		for _, vv := range v {
			injectPromptInto(vv, suffix)
		}
	}
	return obj
}

func shadowTools(obj any, extra map[string]any) any {
	switch v := obj.(type) {
	case map[string]any:
		for k, vv := range v {
			if k == "tools" {
				if arr, ok := vv.([]any); ok {
					cp := map[string]any{}
					for ek, ev := range extra {
						cp[ek] = ev
					}
					v[k] = append(arr, cp)
				}
			} else {
				shadowTools(vv, extra)
			}
		}
	case []any:
		for _, vv := range v {
			shadowTools(vv, extra)
		}
	}
	return obj
}

var dropRequestHeaders = map[string]struct{}{
	"host": {}, "content-length": {}, "connection": {}, "keep-alive": {},
	"proxy-authenticate": {}, "proxy-authorization": {}, "te": {},
	"trailers": {}, "transfer-encoding": {}, "upgrade": {},
}

// ── ChaosProxy ───────────────────────────────────────────────────────────

// ChaosProxy — a Go HTTP forward-proxy with declarative chaos rules.
// Streams SSE chunk-by-chunk. The proxy is plaintext-to-agent; it
// upgrades to TLS on the way out if the upstream is `https://`.
type ChaosProxy struct {
	Upstream  string
	Faults    []ProxyFault
	Timeout   time.Duration
	rng       *rand.Rand
	server    *http.Server
	listener  net.Listener
	url       string
	mu        sync.Mutex
	faultHits map[string]int
}

// ChaosProxyOpts — knobs for the proxy.
type ChaosProxyOpts struct {
	Rng     *rand.Rand
	Timeout time.Duration
}

// NewChaosProxy — construct (do not Start).
func NewChaosProxy(upstream string, faults []ProxyFault, opts ChaosProxyOpts) *ChaosProxy {
	p := &ChaosProxy{
		Upstream:  strings.TrimRight(upstream, "/"),
		Faults:    faults,
		Timeout:   opts.Timeout,
		rng:       opts.Rng,
		faultHits: map[string]int{},
	}
	if p.Timeout == 0 {
		p.Timeout = 60 * time.Second
	}
	if p.rng == nil {
		p.rng = rand.New(rand.NewPCG(rand.Uint64(), rand.Uint64()))
	}
	return p
}

// URL — the proxy's listening URL, populated by Start.
func (p *ChaosProxy) URL() string { return p.url }

// FaultHits — snapshot of per-(kind, pathPrefix) fire counts.
func (p *ChaosProxy) FaultHits() map[string]int {
	p.mu.Lock()
	defer p.mu.Unlock()
	out := make(map[string]int, len(p.faultHits))
	for k, v := range p.faultHits {
		out[k] = v
	}
	return out
}

// Start — bind on `host:port`. Port 0 picks a free port.
func (p *ChaosProxy) Start(host string, port int) error {
	if host == "" {
		host = "127.0.0.1"
	}
	addr := fmt.Sprintf("%s:%d", host, port)
	ln, err := net.Listen("tcp", addr)
	if err != nil {
		return err
	}
	p.listener = ln
	p.url = fmt.Sprintf("http://%s", ln.Addr().String())
	p.server = &http.Server{Handler: http.HandlerFunc(p.handle)}
	go func() { _ = p.server.Serve(ln) }()
	return nil
}

// Stop — graceful shutdown.
func (p *ChaosProxy) Stop(ctx context.Context) error {
	if p.server == nil {
		return nil
	}
	return p.server.Shutdown(ctx)
}

func (p *ChaosProxy) matching(path string, kind ProxyFaultKind) []ProxyFault {
	var out []ProxyFault
	for _, f := range p.Faults {
		if f.Kind != kind {
			continue
		}
		if f.PathPrefix != "" && !strings.HasPrefix(path, f.PathPrefix) {
			continue
		}
		if f.Probability >= 1.0 || p.rng.Float64() < f.Probability {
			out = append(out, f)
			key := fmt.Sprintf("%s:%s", kind, f.PathPrefix)
			p.mu.Lock()
			p.faultHits[key]++
			p.mu.Unlock()
		}
	}
	return out
}

func (p *ChaosProxy) handle(w http.ResponseWriter, r *http.Request) {
	path := r.URL.Path
	if r.URL.RawQuery != "" {
		path += "?" + r.URL.RawQuery
	}

	// PRE-FORWARD: delay
	for _, f := range p.matching(r.URL.Path, FaultDelay) {
		ms := float64(f.Ms)
		if f.Jitter > 0 {
			ms = ms * (1.0 + (p.rng.Float64()*2-1)*f.Jitter)
			if ms < 0 {
				ms = 0
			}
		}
		time.Sleep(time.Duration(ms) * time.Millisecond)
	}

	// PRE-FORWARD: inject_status
	if injected := p.matching(r.URL.Path, FaultInjectStatus); len(injected) > 0 {
		f := injected[0]
		w.Header().Set("Content-Type", "text/plain; charset=utf-8")
		w.Header().Set("X-Tape-Chaos", "inject_status")
		w.Header().Set("Content-Length", strconv.Itoa(len(f.Body)))
		w.WriteHeader(f.Status)
		_, _ = io.WriteString(w, f.Body)
		return
	}

	// Forward
	body, _ := io.ReadAll(r.Body)
	r.Body.Close()
	target, err := url.Parse(p.Upstream + path)
	if err != nil {
		http.Error(w, "bad upstream URL", http.StatusInternalServerError)
		return
	}
	upstreamReq, err := http.NewRequestWithContext(r.Context(), r.Method, target.String(), bytes.NewReader(body))
	if err != nil {
		http.Error(w, "upstream request: "+err.Error(), http.StatusInternalServerError)
		return
	}
	for k, vs := range r.Header {
		if _, drop := dropRequestHeaders[strings.ToLower(k)]; drop {
			continue
		}
		for _, v := range vs {
			upstreamReq.Header.Add(k, v)
		}
	}
	client := &http.Client{Timeout: p.Timeout}
	resp, err := client.Do(upstreamReq)
	if err != nil {
		w.Header().Set("X-Tape-Chaos", "upstream-unreachable")
		w.WriteHeader(http.StatusBadGateway)
		_, _ = io.WriteString(w, "upstream unreachable: "+err.Error())
		return
	}
	defer resp.Body.Close()

	ctype := resp.Header.Get("Content-Type")
	switch {
	case strings.HasPrefix(ctype, "text/event-stream"):
		p.replyStream(r.URL.Path, resp, w)
	case strings.HasPrefix(ctype, "application/json"):
		p.replyJSON(r.URL.Path, resp, w)
	default:
		p.replyPassthrough(r.URL.Path, resp, w)
	}
}

func (p *ChaosProxy) writeHeaders(resp *http.Response, w http.ResponseWriter, overrideLen int, extra map[string]string) {
	for k, vs := range resp.Header {
		lk := strings.ToLower(k)
		if lk == "transfer-encoding" || lk == "content-encoding" ||
			lk == "connection" || lk == "keep-alive" {
			continue
		}
		if overrideLen >= 0 && lk == "content-length" {
			continue
		}
		for _, v := range vs {
			w.Header().Add(k, v)
		}
	}
	if overrideLen >= 0 {
		w.Header().Set("Content-Length", strconv.Itoa(overrideLen))
	}
	for k, v := range extra {
		w.Header().Set(k, v)
	}
	w.WriteHeader(resp.StatusCode)
}

func (p *ChaosProxy) replyPassthrough(path string, resp *http.Response, w http.ResponseWriter) {
	data, _ := io.ReadAll(resp.Body)
	drops := p.matching(path, FaultDropConnection)
	p.writeHeaders(resp, w, len(data), map[string]string{"X-Tape-Chaos": "passthrough"})
	if len(drops) > 0 {
		return // headers sent, body intentionally omitted
	}
	_, _ = w.Write(data)
}

func (p *ChaosProxy) replyJSON(path string, resp *http.Response, w http.ResponseWriter) {
	data, _ := io.ReadAll(resp.Body)
	var payload any
	if err := json.Unmarshal(data, &payload); err != nil {
		p.writeHeaders(resp, w, len(data), nil)
		_, _ = w.Write(data)
		return
	}
	var applied []string
	for _, f := range p.matching(path, FaultMangleJSON) {
		payload = setJSONAt(payload, f.JSONPath, f.Replacement)
		applied = append(applied, "mangle_json")
	}
	for _, f := range p.matching(path, FaultInjectPrompt) {
		payload = injectPromptInto(payload, f.Suffix)
		applied = append(applied, "inject_prompt")
	}
	for _, f := range p.matching(path, FaultToolShadow) {
		if f.ExtraTool != nil {
			payload = shadowTools(payload, f.ExtraTool)
			applied = append(applied, "tool_shadow")
		}
	}
	for _, f := range p.matching(path, FaultSchemaDrift) {
		if f.DriftFn != nil {
			func() {
				defer func() { _ = recover() }()
				payload = f.DriftFn(payload)
			}()
			applied = append(applied, "schema_drift")
		}
	}
	drops := p.matching(path, FaultDropConnection)
	newBody, _ := json.Marshal(payload)
	tag := strings.Join(applied, ",")
	if tag == "" {
		tag = "json"
	}
	p.writeHeaders(resp, w, len(newBody), map[string]string{"X-Tape-Chaos": tag})
	if len(drops) > 0 {
		return
	}
	_, _ = w.Write(newBody)
}

func (p *ChaosProxy) replyStream(path string, resp *http.Response, w http.ResponseWriter) {
	truncates := p.matching(path, FaultTruncateStream)
	cutAt := 0
	if len(truncates) > 0 {
		cutAt = truncates[0].AtEvent
		for _, t := range truncates[1:] {
			if t.AtEvent < cutAt {
				cutAt = t.AtEvent
			}
		}
	}
	drops := p.matching(path, FaultDropConnection)
	tag := "sse"
	if len(truncates) > 0 {
		tag = "truncate_stream"
	} else if len(drops) > 0 {
		tag = "drop_connection"
	}
	p.writeHeaders(resp, w, -1, map[string]string{"X-Tape-Chaos": tag})

	flusher, _ := w.(http.Flusher)
	buf := bytes.Buffer{}
	chunk := make([]byte, 2048)
	eventCount := 0
	for {
		n, err := resp.Body.Read(chunk)
		if n > 0 {
			buf.Write(chunk[:n])
			for {
				idx := bytes.Index(buf.Bytes(), []byte("\n\n"))
				if idx < 0 {
					break
				}
				evt := buf.Next(idx + 2)
				eventCount++
				if _, werr := w.Write(evt); werr != nil {
					return
				}
				if flusher != nil {
					flusher.Flush()
				}
				if cutAt > 0 && eventCount >= cutAt {
					return
				}
			}
			if len(drops) > 0 && eventCount >= 1 {
				return
			}
		}
		if err != nil {
			return
		}
	}
}

// ── convenience constructors ─────────────────────────────────────────────

// ModelProxy — a ChaosProxy tuned for an LLM provider's base URL.
func ModelProxy(upstream string, faults []ProxyFault, opts ChaosProxyOpts) *ChaosProxy {
	return NewChaosProxy(upstream, faults, opts)
}

// MCPProxy — a ChaosProxy tuned for an MCP server's HTTP/SSE endpoint.
func MCPProxy(upstream string, faults []ProxyFault, opts ChaosProxyOpts) *ChaosProxy {
	return NewChaosProxy(upstream, faults, opts)
}
